File: frontend/src/types/predictive.ts
export type WarningLevel = 'INFO' | 'WATCH' | 'ELEVATED' | 'HIGH';
export type LifecycleStatus = 'NEW' | 'ACTIVE' | 'ACKNOWLEDGED' | 'EXPIRED' | 'DISMISSED';
export type TrendStatus = 'INCREASING' | 'STABLE' | 'DECREASING' | 'VOLATILE';

export interface ActivityTimeSeriesPoint {
  timestamp: string;
  actual_count: number;
  expected_count: number;
  lower_bound: number;
  upper_bound: number;
  is_spike: boolean;
  is_drop: boolean;
}

export interface ActivityTimeSeriesResponse {
  target_type: string;
  target_id: string;
  window_minutes: number;
  points: ActivityTimeSeriesPoint[];
  trend: TrendStatus;
  trend_slope: number;
}

export interface ForecastDriverItem {
  driver: string;
  description: string;
}

export interface ForecastCounterSignalItem {
  signal: string;
  description: string;
}

export interface ForecastResponse {
  target_type: string;
  target_id: string;
  forecast_horizon_minutes: number;
  forecast_level: 'LOW' | 'NORMAL' | 'ELEVATED' | 'HIGH' | 'UNKNOWN';
  expected_activity_count: number;
  expected_range_min: number;
  expected_range_max: number;
  forecast_risk_score: number;
  current_risk_score: number;
  confidence: number;
  data_quality_score: number;
  status: 'COMPLETED' | 'INSUFFICIENT_DATA' | 'INFRASTRUCTURE_DEGRADED';
  reasons: ForecastDriverItem[];
  counter_signals: ForecastCounterSignalItem[];
  model_version: string;
}

export interface EarlyWarning {
  id: number;
  warning_id: string;
  warning_level: WarningLevel;
  lifecycle_status: LifecycleStatus;
  camera_id?: string;
  zone_id?: string;
  zone_name?: string;
  site_id?: string;
  forecast_risk_score: number;
  current_risk_score: number;
  confidence: number;
  data_quality_score: number;
  current_activity_count: number;
  baseline_expected_count: number;
  deviation_percent: number;
  trend: TrendStatus;
  forecast_horizon_minutes: number;
  reasons: ForecastDriverItem[];
  counter_signals: ForecastCounterSignalItem[];
  acknowledged_by?: string;
  acknowledged_at?: string;
  notes?: string;
  expires_at: string;
  created_at: string;
  updated_at: string;
}

export interface HotspotZone {
  zone_id: string;
  name: string;
  camera_id: string;
  bop_site: string;
  latitude: number;
  longitude: number;
  hotspot_level: 'NORMAL' | 'WATCH' | 'ELEVATED' | 'HIGH';
  activity_density: number;
  current_activity: number;
  baseline_activity: number;
  deviation_percent: number;
  risk_score: number;
  contributing_factors: string[];
}

export interface RecommendedAttentionItem {
  camera_id: string;
  camera_name: string;
  bop_site: string;
  priority: 'CRITICAL' | 'HIGH' | 'MEDIUM' | 'LOW';
  reason: string;
  risk_score: number;
  trend: TrendStatus;
}

export interface RecommendedAttentionResponse {
  generated_at: string;
  recommendations: RecommendedAttentionItem[];
}

export interface BaselineShift {
  id: number;
  shift_id: string;
  camera_id: string;
  zone_id?: string;
  hour_of_day: number;
  old_baseline_value: number;
  new_observed_value: number;
  deviation_percent: number;
  status: 'REVIEW_REQUIRED' | 'APPROVED' | 'REJECTED';
  detected_at: string;
  reviewed_by?: string;
  reviewed_at?: string;
  notes?: string;
}

export interface ModelHealth {
  id: number;
  model_name: string;
  model_version: string;
  status: string;
  last_trained_at: string;
  training_data_points: number;
  historical_days: number;
  data_quality_score: number;
  mae_score: number;
  rmse_score: number;
  confidence_avg: number;
  updated_at: string;
}

File: backend/app/schemas/predictive.py
from pydantic import BaseModel, Field
from typing import List, Optional, Dict, Any
from datetime import datetime

class ActivityTimeSeriesPoint(BaseModel):
    timestamp: datetime
    actual_count: int
    expected_count: float
    lower_bound: float
    upper_bound: float
    is_spike: bool = False
    is_drop: bool = False

class ActivityTimeSeriesResponse(BaseModel):
    target_type: str # "camera", "zone", "site"
    target_id: str
    window_minutes: int
    points: List[ActivityTimeSeriesPoint] = []
    trend: str # "INCREASING", "STABLE", "DECREASING", "VOLATILE"
    trend_slope: float = 0.0

class ForecastDriverItem(BaseModel):
    driver: str
    description: str

class ForecastCounterSignalItem(BaseModel):
    signal: str
    description: str

class ForecastResponse(BaseModel):
    target_type: str # "camera", "zone", "site"
    target_id: str
    forecast_horizon_minutes: int = 60
    forecast_level: str # "LOW", "NORMAL", "ELEVATED", "HIGH"
    expected_activity_count: float
    expected_range_min: float
    expected_range_max: float
    forecast_risk_score: int
    current_risk_score: int
    confidence: float # 0.0 to 1.0
    data_quality_score: float # 0.0 to 1.0
    status: str = "COMPLETED" # "COMPLETED", "INSUFFICIENT_DATA", "INFRASTRUCTURE_DEGRADED"
    reasons: List[ForecastDriverItem] = []
    counter_signals: List[ForecastCounterSignalItem] = []
    model_version: str = "1.0.0"

class EarlyWarningResponse(BaseModel):
    id: int
    warning_id: str
    warning_level: str
    lifecycle_status: str
    camera_id: Optional[str] = None
    zone_id: Optional[str] = None
    zone_name: Optional[str] = None
    site_id: Optional[str] = None
    forecast_risk_score: int
    current_risk_score: int
    confidence: float
    data_quality_score: float
    current_activity_count: int
    baseline_expected_count: float
    deviation_percent: float
    trend: str
    forecast_horizon_minutes: int
    reasons: List[ForecastDriverItem] = []
    counter_signals: List[ForecastCounterSignalItem] = []
    acknowledged_by: Optional[str] = None
    acknowledged_at: Optional[datetime] = None
    notes: Optional[str] = None
    expires_at: datetime
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True

class EarlyWarningUpdate(BaseModel):
    lifecycle_status: Optional[str] = None
    notes: Optional[str] = None
    actor_username: Optional[str] = "operator"

class HotspotZoneResponse(BaseModel):
    zone_id: str
    name: str
    camera_id: str
    bop_site: str
    latitude: float
    longitude: float
    hotspot_level: str # "NORMAL", "WATCH", "ELEVATED", "HIGH"
    activity_density: float # events per km² or relative index
    current_activity: int
    baseline_activity: float
    deviation_percent: float
    risk_score: int
    contributing_factors: List[str] = []

class RecommendedAttentionItem(BaseModel):
    camera_id: str
    camera_name: str
    bop_site: str
    priority: str # "CRITICAL", "HIGH", "MEDIUM", "LOW"
    reason: str
    risk_score: int
    trend: str

class RecommendedAttentionResponse(BaseModel):
    generated_at: datetime
    recommendations: List[RecommendedAttentionItem] = []

class BaselineShiftResponse(BaseModel):
    id: int
    shift_id: str
    camera_id: str
    zone_id: Optional[str] = None
    hour_of_day: int
    old_baseline_value: float
    new_observed_value: float
    deviation_percent: float
    status: str
    detected_at: datetime
    reviewed_by: Optional[str] = None
    reviewed_at: Optional[datetime] = None
    notes: Optional[str] = None

    class Config:
        from_attributes = True

class BaselineShiftApprove(BaseModel):
    approve: bool
    notes: Optional[str] = None
    reviewed_by: Optional[str] = "admin"

class ModelHealthResponse(BaseModel):
    id: int
    model_name: str
    model_version: str
    status: str
    last_trained_at: datetime
    training_data_points: int
    historical_days: int
    data_quality_score: float
    mae_score: float
    rmse_score: float
    confidence_avg: float
    updated_at: datetime

    class Config:
        from_attributes = True

class PredictionFeedbackCreate(BaseModel):
    warning_id: str
    feedback_type: str = Field(..., description="USEFUL_FORECAST, FALSE_PREDICTION, INCONCLUSIVE")
    notes: Optional[str] = None
    operator_username: Optional[str] = "operator"

class PredictionFeedbackResponse(BaseModel):
    id: int
    feedback_id: str
    warning_id: str
    feedback_type: str
    operator_username: str
    notes: Optional[str] = None
    created_at: datetime

    class Config:
        from_attributes = True
